node/libs/concurrency/src/sync/prunable_queue.rs

```diff
@@ -0,0 +1,148 @@
+) -> (Sender<T, U>, Receiver<T, U>) {
+    let queue: Mutex<VecDeque<(T, oneshot::Sender<U>)>> = Mutex::new(VecDeque::new());
+    // Internal signaling, to enable waiting on the receiver side for new items.
+    let (send, recv) = watch::channel(false);
+
+    let shared = Arc::new(Shared {
+        queue,
+        has_items_send: send,
+    });
+
+    let sender = Sender {
+        shared: shared.clone(),
+        pruning_predicate: pruning_predicate,
+    };
+
+    let receiver = Receiver {
+        shared: shared.clone(),
+        has_items_recv: recv,
+    };
+
+    return (sender, receiver);
+}
+
+pub struct Shared<T, U> {
+    queue: Mutex<VecDeque<(T, oneshot::Sender<U>)>>,
+    has_items_send: watch::Sender<bool>,
+}
+
+pub struct Sender<T, U> {
+    shared: Arc<Shared<T, U>>,
+    pruning_predicate: Box<dyn Sync + Send + Fn(&T, &T) -> bool>,
+}
+
+impl<T, U> Sender<T, U> {
+    pub async fn enqueue(&self, item: T) -> oneshot::Receiver<U> {
+        // Create oneshot channel for returning result asynchronously.
+        let (res_send, res_recv) = oneshot::channel();
+
+        let mut queue = self.shared.queue.lock().await;
+        queue.retain(|existing_item| !(self.pruning_predicate)(&existing_item.0, &item));
+        queue.push_back((item, res_send));
+
+        // Ignore sending error.
+        let _ = self.shared.has_items_send.send(true);
+
+        res_recv
+    }
+}
+
+pub struct Receiver<T, U> {
+    shared: Arc<Shared<T, U>>,
+    has_items_recv: watch::Receiver<bool>,
+}
+
+impl<T, U> Receiver<T, U> {
+    pub async fn dequeue(&mut self, ctx: &Ctx) -> ctx::OrCanceled<(T, oneshot::Sender<U>)> {
+        sync::wait_for(ctx, &mut self.has_items_recv, |has_items| *has_items).await?;
+        let mut queue = self.shared.queue.lock().await;
+        // `None` is unexpected because we waited for new items.
+        let item = queue.pop_front().unwrap();
+
+        if queue.len() == 0 {
+            // Send error is unexpected because `self` holds the receiver.
+            self.shared.has_items_send.send(false).unwrap();
+        }
+
+        Ok(item)
+    }
+}
+
+// Test scenario:
+// Enqueue two sets of 0..1000 items, in conjunction, while pruning
+// so that only one 0..1000 set is expected to remain in queue.
+// Then, dequeue to assert the queue's content.
+#[tokio::test]
+async fn test_prunable_queue() {
+    use tokio::time::{timeout, Duration};
+
+    #[derive(Debug, Clone)]
+    struct ItemType(usize, usize);
+
+    let ctx = ctx::test_root(&ctx::RealClock);
+
+    let (sender, mut receiver) = new(Box::new(|a: &ItemType, b: &ItemType| a.0 != b.0));
+
+    let sender1 = Arc::new(sender);
+    let sender2 = sender1.clone();
+
+    let handle1 = tokio::spawn(async move {
+        let set = 1;
+        let items = (0..1000).map(|i| ItemType(i, set));
+        for item in items {
+            let _ = sender1.enqueue(item).await;
+            tokio::task::yield_now().await;
```

Comment:
I'd suggest to manually implement a deterministic random interleaving, since it is feasible here (rather than rely on yield_now).

Reply:
I left this code because I planned to further implement something non-deterministic, but it's probably a bad idea. 
I'll do something more reasonably-deterministic as you suggested, without relying on `yield_now`.

Reply:
Test code changes here: https://github.com/matter-labs/era-consensus/pull/56/commits/020f25b2049f9f4bb2dbbddc9a0e876eadaf5389

It doesn't involve anything sophisticated, just covers the basic features.